### PR TITLE
Misc cleanups in ink! codegen

### PR DESCRIPTION
**Pull request by Robbepop**
Work carved out of https://github.com/paritytech/ink/pull/665 in an attempt to make its diff more manageable.

- Some macro hygiene improvements.
- Modernization of some codegen parts.
- Simplification of codegen parts.

### CODE REVIEW

File: crates/lang/codegen/src/generator/metadata.rs [hunk 1]
```diff
@@ -84,16 +84,16 @@ impl Metadata<'_> {
         quote! {
             ::ink_metadata::ContractSpec::new()
                 .constructors([
-                    #(#constructors ,)*
+                    #( #constructors ),*
```
Comment (HCastano) [inline crates/lang/codegen/src/generator/metadata.rs:87]: Does this actually make a difference as far as `quote` is concerned, or is it an aesthetic preference?

File: crates/lang/codegen/src/generator/metadata.rs [hunk 2]
```diff
@@ -270,7 +275,7 @@ impl Metadata<'_> {
         match ret_ty {
             None => {
                 quote! {
-                    ::ink_metadata::ReturnTypeSpec::new(None)
+                    ::ink_metadata::ReturnTypeSpec::new(::core::option::Option::None)
```
Comment (HCastano) [inline crates/lang/codegen/src/generator/metadata.rs:278]: I feel like we should be able to do all these path clean-ups with a Bash script

File: crates/lang/codegen/src/generator/mod.rs [hunk 3]
```diff
@@ -12,6 +12,20 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+/// Implements `AsRef` for a code generator type.
+///
+/// Code generators always have a shared `contract` reference to the contract.
+/// They need to implement this trait in order to use other code generators.
+macro_rules! impl_as_ref_for_generator {
```
Comment (HCastano) [inline crates/lang/codegen/src/generator/mod.rs:19]: nit: I'd move this to after the imports

File: crates/lang/codegen/src/generator/storage.rs [hunk 4]
```diff
@@ -30,24 +30,18 @@ use syn::spanned::Spanned as _;
 pub struct Storage<'a> {
     contract: &'a ir::Contract,
 }
-
-impl<'a> AsRef<ir::Contract> for Storage<'_> {
-    fn as_ref(&self) -> &ir::Contract {
-        self.contract
-    }
-}
+impl_as_ref_for_generator!(Storage);
```
Comment (HCastano) [inline crates/lang/codegen/src/generator/storage.rs:33]:
```suggestion

impl_as_ref_for_generator!(Storage);
```

File: crates/lang/codegen/src/generator/storage.rs [hunk 5]
```diff
@@ -30,24 +30,18 @@ use syn::spanned::Spanned as _;
...
         let access_env_impls = self.generate_access_env_trait_impls();
         let storage_struct = self.generate_storage_struct();
-        let use_emit_event = if self.contract.module().events().next().is_some() {
-            // Required to allow for `self.env().emit_event(..)` in messages and constructors.
-            Some(quote! { use ::ink_lang::EmitEvent as _; })
-        } else {
-            None
-        };
+        let use_emit_event =
+            self.contract.module().events().next().is_some().then(|| {
```
Comment (HCastano) [inline crates/lang/codegen/src/generator/storage.rs:41]: Nice 😛

**Review by HCastano — APPROVED** [review]

File: crates/lang/codegen/src/generator/metadata.rs (hunk 1 above)
Comment (Robbepop) [reply crates/lang/codegen/src/generator/metadata.rs:87]: It does make a difference if the list of constructors is not empty.
Before: `.constructors([ first, second, ])`
Now: `.constructors([ first, second ])`
Note the missing comma at the end.

File: crates/lang/codegen/src/generator/metadata.rs (hunk 2 above)
Comment (Robbepop) [reply crates/lang/codegen/src/generator/metadata.rs:278]: I hope we won't be trusting a bash script for transformations like these on our code. :sweat_smile:

File: crates/lang/codegen/src/generator/mod.rs (hunk 3 above)
Comment (Robbepop) [reply crates/lang/codegen/src/generator/mod.rs:19]: then it won't work ... Rust's macro system needs this ordering unfortunately.

File: crates/lang/codegen/src/generator/storage.rs (hunk 5 above)
Comment (Robbepop) [reply crates/lang/codegen/src/generator/storage.rs:41]: I really like this new API. ^^
Another nice one is `array::map`.

File: crates/lang/codegen/src/generator/metadata.rs (hunk 1 above)
Comment (HCastano) [reply crates/lang/codegen/src/generator/metadata.rs:87]: 🤦‍♂️ didn't notice that while skimming through

File: crates/lang/codegen/src/generator/mod.rs (hunk 3 above)
Comment (HCastano) [reply crates/lang/codegen/src/generator/mod.rs:19]: I feel like I knew that deep down, but TIL

**Pull request merged**